### DISCUSSION TITLE
When in IndexField to return keys or not

### DISCUSSION THREAD

**Discussion by antonincms** [kubernetes-sigs/kubebuilder#2915] [Support] [answered]
I was searching on explanation on how to index fields using the cache since I had an issue where I create an object, in the next reconcile loop was not finding it and I was thinking that cache could be the reason. 

In the book there is one main reference to that [here](https://book.kubebuilder.io/reference/watching-resources/externally-managed.html) :
```
    if err := mgr.GetFieldIndexer().IndexField(context.Background(), &appsv1.ConfigDeployment{}, configMapField, func(rawObj client.Object) []string {
        // Extract the ConfigMap name from the ConfigDeployment Spec, if one is provided
        configDeployment := rawObj.(*appsv1.ConfigDeployment)
        if configDeployment.Spec.ConfigMap == "" {
            return nil
        }
        return []string{configDeployment.Spec.ConfigMap}
    }); err != nil {
        return err
    }
``` 

But I was wondering why if the ConfigMap is empty nothing is returned, what is the difference between doing that and simply always returning `[]string{configDeployment.Spec.ConfigMap}` ?

**Comment by camilamacedo86** [comment]
HI @antonincms,

If you still needing helping with this one, could you please clarifies your question:

> But I was wondering why if the ConfigMap is empty nothing is returned, what is the difference between doing that and simply always returning []string{configDeployment.Spec.ConfigMap} ?

**Reply by antonincms** [reply]
Hello @camilamacedo86 , 

I was wondering what is the difference in behavior between the previous snippet and that one :  
```
    if err := mgr.GetFieldIndexer().IndexField(context.Background(), &appsv1.ConfigDeployment{}, configMapField, func(rawObj client.Object) []string {
        // Extract the ConfigMap name from the ConfigDeployment Spec
        configDeployment := rawObj.(*appsv1.ConfigDeployment)
        return []string{configDeployment.Spec.ConfigMap}
    }); err != nil {
        return err
    }
```

I was trying to understan fieldsIndexers better but I did not get in which case it made sense to return nothing or several fields.

**Reply by camilamacedo86** [reply, marked as answer]
Hi @antonincms,

The purpose of indexing in the cache is to allow quick lookups based on certain fields' values. An indexed field will have a mapping from values of that field to the objects that have those values.

Considering your two scenarios:

Returning nil when ConfigMap is empty:

```go
if err := mgr.GetFieldIndexer().IndexField(context.Background(), &appsv1.ConfigDeployment{}, configMapField, func(rawObj client.Object) []string {
    // Extract the ConfigMap name from the ConfigDeployment Spec, if one is provided
    configDeployment := rawObj.(*appsv1.ConfigDeployment)
    if configDeployment.Spec.ConfigMap == "" {
        return nil
    }
    return []string{configDeployment.Spec.ConfigMap}
}); err != nil {
    return err
}
```

Always returning the value of ConfigMap, even if it's empty:

```go
if err := mgr.GetFieldIndexer().IndexField(context.Background(), &appsv1.ConfigDeployment{}, configMapField, func(rawObj client.Object) []string {
    // Extract the ConfigMap name from the ConfigDeployment Spec
    configDeployment := rawObj.(*appsv1.ConfigDeployment)
    return []string{configDeployment.Spec.ConfigMap}
}); err != nil {
    return err
}
```
The difference between these two scenarios:

- In the first scenario, if the ConfigMap field is empty (""), then this object will not be indexed by the ConfigMap field at all. This means that a query trying to find ConfigDeployment objects by an empty ConfigMap field will return nothing.
- In the second scenario, every ConfigDeployment object will be indexed by the ConfigMap field, even if it's empty. A query trying to find ConfigDeployment objects by an empty ConfigMap field will return all the ConfigDeployment objects that have an empty ConfigMap field.

Deciding between these approaches depends on the use-case:

- If there's no meaningful reason to query for objects with an empty ConfigMap field, and you want to optimize the cache's storage, then the first scenario is more appropriate.
- If, however, you have a reason to look up objects that specifically have an empty ConfigMap field, then the second scenario is more appropriate.

I hope that clarifies the difference!

**Reply by antonincms** [reply]
Thanks a lot for the detailed answer, it is clearer now.

**Comment by camilamacedo86** [comment]
I am closing this one as answered
however, if you need please feel free to reopen
